gen/bind_generator.jl

```diff
@@ -0,0 +1,147 @@
+If a `String`, the string is used as the message in an `error()` call, otherwise the
+expression is used as-is. Note that the expression may refer to any arguments by name.
+
+The declared return type in the function-like signature must be the return type of the C
+function, and the Julia return type is inferred as one of the following possibilities:
+
+1. If `ReturnType === :Herr`, the Julia function returns `nothing` and the C return is
+   used only in error checking.
+
+2. If `ReturnType === :Htri`, the Julia function returns a boolean indicating whether
+   the return value of the C function was zero (`false`) or positive (`true`).
+
+3. Otherwise, the C function return value is returned from the Julia function.
+
+Furthermore, the C return value is interpreted to automatically generate error checks
+in most cases.
+
+1. If `ReturnType === :Herr` or `ReturnType === :Htri`, an error is raised when the return
+   value is negative.
+
+2. If `ReturnType === :Haddr` or `ReturnType === :Hsize`, an error is raised when the
+   return value is equivalent to `-1 % Haddr` and `-1 % Hsize`, respectively.
+
+3. For all other return types, it is assumed a negative value indicates error.
+
+It is assumed that the HDF library names are given in global constants named `libhdf5`
+and `libhdf5_hl`. The former is used for all `ccall`s, except if the C library name begins
+with "H5DO" or "H5TB" then the latter library is used.
+"""
+macro bind(sig::Expr, err::Union{String,Expr})
+    sig.head === :(::) || error("return type required on function signature")
+
+    # Pull apart return-type and rest of function declaration
+    rettype = sig.args[2]::Union{Symbol,Expr}
+    funcsig = sig.args[1]
+    isexpr(funcsig, :call) || error("expected function-like expression, found `", funcsig, "`")
+    funcsig = funcsig::Expr
+
+    # Extract function name and argument list
+    jlfuncname = funcsig.args[1]::Symbol
+    funcargs = funcsig.args[2:end]
+
+    # Pull apart argument names and types
+    args = Vector{Symbol}()
+    argt = Vector{Union{Expr,Symbol}}()
+    for ii in 1:length(funcargs)
+        argex = funcargs[ii]
+        if !isexpr(argex, :(::)) || !(argex.args[1] isa Symbol)
+            error("expected `name::type` expression in argument ", ii, ", got ", funcargs[ii])
+        end
+        push!(args, argex.args[1])
+        push!(argt, argex.args[2])
+    end
+
+    # Translate the C function name to a local equivalent
+    if haskey(bind_exceptions, jlfuncname)
+        cfuncname = bind_exceptions[jlfuncname]
+    else
+        # turn e.g. h5f_close into H5Fclose
+        prefix, rest = split(string(jlfuncname), "_", limit = 2)
+        cfuncname = Symbol(uppercase(prefix), rest)
+    end
+
+    # Determine the underlying C library to call
+    lib = startswith(string(cfuncname), r"H5(DO|TB)") ? :libhdf5_hl : :libhdf5
+
+    # Now start building up the full expression:
+    statsym = Symbol("#status#") # not using gensym() to have stable naming
```

Comment:
Does this really need to be `var"#status"`, why not just leave it as, say, `r` or `ret` ?

Reply:
The idea was to make a naming clash with the function arguments unlikely, since we're not actually relying on macro hygiene to ensure the internal and externally-provided symbols are unique.

Reply:
Hmm, I see. It would be pretty unlikely someone do name an argument `ret`, but I suppose it "could" happen.  I guess I'm just a little apprehensive since it isn't very idiomatic.

Reply:
We could change it to something more normal yet still exotic — `__status_` or something.

Reply:
I'm on the fence, we could just keep as is  or name it `status` where it would be pretty straightforward  to catch errors if people commit the wrong symbol names.